blob/service.go

```diff
@@ -0,0 +1,308 @@
+	"github.com/celestiaorg/celestia-node/state"
+)
+
+// TODO(@vgonkivs): remove after bumping celestia-app
+// defaultMinGasPrice is the default min gas price.
+const defaultMinGasPrice = 0.001
+
+var (
+	log = logging.Logger("blob")
+
+	ErrBlobNotFound = errors.New("blob: not found")
+)
+
+type Service struct {
+	// accessor dials the given celestia-core endpoint to submit blobs.
+	accessor *state.CoreAccessor
+	// sGetter retrieves the EDS to fetch all shares from the requested header.
+	sGetter share.Getter
+	// hStore allows to get the requested header from the local storage.
+	hStore libhead.Store[*header.ExtendedHeader]
+	// headGetter gets current network head.
+	headGetter libhead.Head[*header.ExtendedHeader]
+}
+
+func NewService(
+	state *state.CoreAccessor,
+	getter share.Getter,
+	hStore libhead.Store[*header.ExtendedHeader],
+	headGetter libhead.Head[*header.ExtendedHeader],
+) *Service {
+	return &Service{
+		accessor:   state,
+		sGetter:    getter,
+		hStore:     hStore,
+		headGetter: headGetter,
+	}
+}
+
+// Submit sends PFB transaction and reports the height it landed on.
+// Allows sending multiple Blobs atomically synchronously.
+// Uses default wallet registered on the Node.
+func (s *Service) Submit(ctx context.Context, blobs ...*Blob) (uint64, error) {
+	log.Debugw("submitting blobs", "amount", len(blobs))
+
+	var (
+		gasLimit = estimateGas(blobs...)
+		fee      = int64(defaultMinGasPrice * float64(gasLimit))
+		b        = make([]*apptypes.Blob, len(blobs))
+	)
+
+	for i, blob := range blobs {
+		b[i] = &blob.Blob
+	}
+
+	resp, err := s.accessor.SubmitPayForBlob(ctx, types.NewInt(fee), gasLimit, b...)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(resp.Height), nil
+}
+
+// Get retrieves all the blobs for given namespaces at the given height by commitment.
+func (s *Service) Get(ctx context.Context, height uint64, nID namespace.ID, commitment Commitment) (*Blob, error) {
+	blob, _, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return blob, nil
+}
+
+// GetProof retrieves all blobs in the given namespaces at the given height by commitment
+// and returns their Proof.
+func (s *Service) GetProof(
+	ctx context.Context,
+	height uint64,
+	nID namespace.ID,
+	commitment Commitment,
+) (*Proof, error) {
+	_, proof, err := s.getByCommitment(ctx, height, nID, commitment)
+	if err != nil {
+		return nil, err
+	}
+	return proof, nil
+}
+
+// GetAll returns all blobs under the given namespaces at the given height.
+// GetAll can return blobs and an error in case if some requests failed.
+func (s *Service) GetAll(ctx context.Context, height uint64, nIDs ...namespace.ID) ([]*Blob, error) {
+	header, err := s.getByHeight(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		blobs     = make([]*Blob, 0)
+		ch        = make(chan []*Blob)
+		resultErr = make([]error, len(nIDs))
+	)
+
+	for i, nID := range nIDs {
+		i := i
+		nID := nID
+		go func() {
```

Comment:
pls extract into func

Reply:
why? it invokes getBlobs and then handles the result. p2p.Exchange has pretty same logic in one place

Reply:
reworked